### DISCUSSION TITLE
Nested RequestContext and detect entity from Identity Map

### DISCUSSION THREAD

**Discussion by OnikurYH** [mikro-orm/mikro-orm#3579] [Q&A] [answered]
Hello! I am new to using MikroORM from other ORM libraries and got some questions when using it. Was trying to find it on the internet but still cannot find the solution.

- - -

Current I using version 5.3.0, I see the [document](https://mikro-orm.io/docs/identity-map#userequestcontext-decorator) said `UseRequestContext` should be used only on the top-level methods and it should not be nested, may I know what is the reason why it should not be nested? I am using NestJS so it auto-applied RequestContext, but also will call some async function inside the controller -> service so we want it on another request context.
**Code example**
```ts
class MyService {
    @UseRequestContext()
    async updateSomething() {
        // ... something done
        this.somethingRunInBackground().catch(() => /* Error handling */);
        // ... something to continue working on
    }

    @UseRequestContext()
    async somethingRunInBackground() {
        // Something will do in the background, and it will do find and save, but don't want to affect the outer scope
    }
}
```

- - -

Another question is can we manually detect the entity after it get from the repository? I know there is an option `disableIdentityMap ` on the repo `find*` function but seems there is no method to detect it later?

Thanks!

**Comment by B4nan** [comment, marked as answer]
> may I know what is the reason why it should not be nested?

Because it does not make any sense, the whole point of having the context is not to mix it with other contexts - which is exactly what you do in that code snippet. Moreover, you should not put it on service methods, this belongs to the controller if something.

>  I am using NestJS 

The most important thing is - you should not use this for regular requests! This is only for a specific use case like CRON jobs or event handlers. The context is automatically managed by the `RequestContext` helper, registered via middleware in the `@mikro-orm/nest` package. You just use the EM and repositories from DI container (the global ones) and thanks to this smart middleware it will automatically hook the correct context for you.

https://mikro-orm.io/docs/identity-map#how-does-requestcontext-helper-work

> Another question is can we manually detect the entity after it get from the repository? 

You can merge a detached entity, if thats what you mean.

**Reply by OnikurYH** [reply]
Thanks for your quick reply B4nan!
Got it from points 1 & 2.
About the detached entity I was some typos in my question, actually, I want to know if is there any way can detach the entity from the identity map (not managed) when the entity was managed? Moreover, I was try to find the method `em.registerManaged()` from this [document](https://mikro-orm.io/docs/entity-manager#disabling-identity-map-and-change-set-tracking), but when I try to use on my code it is undefined, is this not exists anymore in v5.3 or I missed something? (The `em` I get from the type `MikroORM<PostgreSqlDriver>`)

**Reply by B4nan** [reply]
> Moreover, I was try to find the method em.registerManaged() from this [document](https://mikro-orm.io/docs/entity-manager#disabling-identity-map-and-change-set-tracking),

Oh good catch, the method is in the UoW (so `em.getUnitOfWork().registerManager()`), the EM API you should be using is `em.merge()` (it uses this method under the hood). 

There is also `em.getUnitOfWork().unsetIdentity()`, but its rather internal API, I am affraid it could have side effects - if you detach entity that was referenced from some owning side, it would probably produce queries to "unwire" this relation (once you call flush). 

Whats the use case? Why would you load some entity into context and remove it afterwards?

**Reply by OnikurYH** [reply]
For the detach because it depends on the nested `RequestContext`question, sometimes want to manage the entity by my owned codebase.
I know the `RequestContext` should be used on some use cases like CRON jobs or event handlers and it will automatically be managed in Nestjs, is there any way can create another context without using events or a job queue from the same request? Now is a case when calling an HTTP request, and wanting to do something in the background without blocking, but the background will do some query or save data to the database, so it may also affect the outer scope entity if they are the same, but that is not what I want.

**Reply by B4nan** [reply]
> is there any way can create another context 

`em.fork()` is what creates new context, you can use that explicitly. But you will need to work with the fork, not with any global repositories/services from a DI container - thats the difference from `@UseRequestContext` decorator - it will allow you to use the DI services transparently.

Also, all what this decorator does is using `RequestContext.createAsync` around the function body, you could implement your own decorator that does something similar, but fits more into your needs. It all depends on what is the exact use case.

> and wanting to do something in the background without blocking, but the background will do some query or save data to the database, so it may also affect the outer scope entity if they are the same, but that is not what I want.

Makes sense, yeah, I would use `em.fork()` here and work with that fork explicitly, that is the safest way. If you create a fork manually this way, it will never use the context created by some middleware (in other words, it will ignore the `RequestContext` helper). Or if you want to use the DI services, you might as well use the `RequestContext.createAsync()` explicitly at the call site - instead of calling a method decorated with `@UseRequestContext`, you would create this context earlier and explicitly.

**Reply by anthonylau** [reply]
> you might as well use the `RequestContext.createAsync()` explicitly at the call site - instead of calling a method decorated with `@UseRequestContext`

I modified the example by @OnikurYH a bit. 

If it is calling from `MyServiceA` to `MyServiceB`. I see that decorate both top level methods instead of adding `RequestContext.createAsync()` in `MyServiceA` can ensure `MyServiceB` methods will be in correct context.

Is there any side effect of this nested context method calls? Seems to be fine if we can use `RequestContext.createAsync()` inside a method with the `@UseRequestContext()`, but I'm not sure the underlying implementation.

```ts
class MyServiceA {
    @UseRequestContext()
    async updateSomething() {
        // ... something done
        this.myServiceB.somethingRunInBackground().catch(() => /* Error handling */);
        // ... something to continue working on
    }
}

class MyServiceB {
    @UseRequestContext()
    async somethingRunInBackground() {
        // Something will do in the background, and it will do find and save, but don't want to affect the outer scope
    }
}
```

**Reply by B4nan** [reply]
It does not ensure correct context, it ensures new context, which is often not what you want. If it is, feel free to use it. My suggestion was to drop the decorator from both places and use the RQ helper only on one of them.

**Reply by OnikurYH** [reply]
Thank you @B4nan and @anthonylau I think that solved my question! OwO
